crates/extensions/tedge_timer_ext/src/builder.rs

```diff
@@ -0,0 +1,202 @@
+use crate::actor::TimerActor;
+use crate::actor::TimerId;
+use crate::SetTimeout;
+use crate::Timeout;
+use async_trait::async_trait;
+use std::collections::HashMap;
+use std::convert::Infallible;
+use std::sync::Arc;
+use tedge_actors::Actor;
+use tedge_actors::Builder;
+use tedge_actors::ChannelError;
+use tedge_actors::DynSender;
+use tedge_actors::Message;
+use tedge_actors::MessageBoxPlug;
+use tedge_actors::MessageBoxSocket;
+use tedge_actors::NoConfig;
+use tedge_actors::Sender;
+use tedge_actors::ServiceMessageBoxBuilder;
+use tokio::sync::Mutex;
+
+pub struct TimerActorBuilder {
+    box_builder: ServiceMessageBoxBuilder<SetTimeout<TimerId>, Timeout<TimerId>>,
+}
+
+impl Default for TimerActorBuilder {
+    fn default() -> Self {
+        TimerActorBuilder {
+            box_builder: ServiceMessageBoxBuilder::new("Timer", 16),
+        }
+    }
+}
+
+impl Builder<(TimerActor, <TimerActor as Actor>::MessageBox)> for TimerActorBuilder {
+    type Error = Infallible;
+
+    fn try_build(self) -> Result<(TimerActor, <TimerActor as Actor>::MessageBox), Self::Error> {
+        Ok(self.build())
+    }
+
+    fn build(self) -> (TimerActor, <TimerActor as Actor>::MessageBox) {
+        let actor = TimerActor::default();
+        let actor_box = self.box_builder.build();
+        (actor, actor_box)
+    }
+}
+
+impl<T: Message> MessageBoxSocket<SetTimeout<T>, Timeout<T>, NoConfig> for TimerActorBuilder {
+    fn connect_with(
+        &mut self,
+        peer: &mut impl MessageBoxPlug<SetTimeout<T>, Timeout<T>>,
+        config: NoConfig,
+    ) {
+        let mut adapter = GenericTimerAdapter::new(peer);
+        self.box_builder.connect_with(&mut adapter, config);
+    }
+}
+
+/// A message adapter used by actors to send timer requests with a generic payload `SetTimeout<T>`
+/// and to receive accordingly timer responses with a generic payload `Timeout<T>`,
+/// while the timer actor only handles `usize` payloads.
```

Comment:
I'm sure there must have been a compelling reason for this indirection but curious to know why the actor couldn't cache the generic `event` payload directly in its state? That would have been simpler to follow.

Reply:
> why the actor couldn't cache the generic event payload directly in its state?

Indeed by far simpler. However, all the actors of this actor would have to use the *same* type of payload.

An option that can be considered is to use [`Any`](https://doc.rust-lang.org/std/any/). The caller will have to provide specific `DynSender` implementation to cast back and forth the generic values into `Any` values.